### PR TITLE
Implement vote calculator API and tests.

### PR DESCRIPTION
**Pull request by Oleh-Stefanovych**
Implement basic functionality for a vote calculator. The API has the ability to:
* elect and declare leader nodes
* elect until a single leader is declared.

### CODE REVIEW

File: VoteCalculator/app/vote-calculator/index.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+        for (let votingNode of votingNodes) {
+
+            if (votingNode.votesNumber === maximumVotesNumber) {
+
+                leaders.push(votingNode.node);
+            }
+        }
+
+        return new ElectionOutcome(leaders, maximumVotesNumber);
```
Comment (dkalinin) [inline VoteCalculator/app/vote-calculator/index.js]: Requirement is not met:

> Also, during the submission of votes some node may gain the early superiority in vote count. E.g. with 7 nodes it’s enough to get 4 votes to get elected, so the remaining three votes do not matter. In this case the leader is declared earlier.

File: VoteCalculator/test/voting-node-test/index.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    assert.equal(votingNode.voteForCandidate(new Set()), null,
+        "Voting node returns null if there are no candidates.");
+});
+
+QUnit.test("Voting node should vote for itself if it is the only candidate.", function (assert) {
+
+    const votingNode = new VotingNode(new Node("I am the voter"));
+    assert.equal(votingNode.voteForCandidate(new Set().add(votingNode)), votingNode,
+        "Voting node votesTable for itself if it is the only candidate.");
```
Comment (dkalinin) [inline VoteCalculator/test/voting-node-test/index.js]: Make the elections equitable! One should not vote for himself.

File: VoteCalculator/test/vote-calculator-test/index.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        if (!nodes.has(leader)) {
+
+            isLeadersAreRegistered = false;
+            break;
+        }
+    }
+
+    assert.ok(isLeadersAreRegistered,
+        "The electSingleLeader method returns a leaders that are registered nodes.");
+});
```
Comment (dkalinin) [inline VoteCalculator/test/vote-calculator-test/index.js]: Please test the case when there are 2 leaders.

File: VoteCalculator/app/node/index.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    /**
+     * Overrides method that converts the instance of a current class to a string that is equal
+     * to its _identifier.
+     *
+     * @returns {string}
+     */
+    toString() {
+
+        return this.identifier;
```
Comment (dkalinin) [inline VoteCalculator/app/node/index.js]: This `toString` is pretty similar to String's `toString`.

File: VoteCalculator/test/voting-node-test/index.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    const votingNode = new VotingNode(new Node("I am the voter"));
+
+    const candidates = new Set()
+        .add(new VotingNode(new Node("I am candidate 1")))
+        .add(new VotingNode(new Node("I am candidate 2")))
+        .add(new VotingNode(new Node("I am candidate 3")));
+
+    const elect = votingNode.voteForCandidate(candidates);
+
+    assert.ok(candidates.has(elect), "Voting node votesTable for any of the possible nodes.");
```
Comment (dkalinin) [inline VoteCalculator/test/voting-node-test/index.js]: What is `votesTable`? I can see it in several places.

File: VoteCalculator/test/voting-node-test/index.js (hunk 2 above)
Comment (Oleh-Stefanovych) [reply VoteCalculator/test/voting-node-test/index.js]: But if this node is the only candidate?

File: VoteCalculator/app/vote-calculator/index.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+     * leaders and the number of votes they received.
+     */
+    electLeaders() {
+
+        const votingNodes = new Set();
+        this._nodes.forEach(node => votingNodes.add(new VotingNode(node)));
+
+        let maximumVotesNumber = 0;
+
+        for (let votingNode of votingNodes) {
```
Comment (Oleh-Stefanovych) [inline VoteCalculator/app/vote-calculator/index.js]: @dkalinin, PTAL, may I use for(let .. of ..) instead of forEach(..) here?

**Comment by Oleh-Stefanovych** [comment]
@dkalinin, fixed, PTAL.

File: VoteCalculator/app/vote-calculator/index.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+     * the number of votes, the election process is stopped and the leader is declared.
+     *
+     * @returns {ElectionOutcome} - result of election process. Contains chosen leaders and the
+     * number of votes that they received.
+     */
+    electLeaders() {
+
+        const electionCommission = new ElectionCommission(this._votingNodes);
+
+        this._votingNodes.forEach(function (votingNode) {
```
Comment (Oleh-Stefanovych) [inline VoteCalculator/app/vote-calculator/index.js]: @dkalinin, may I use `for(let .. of ..)` intead of `forEach(..)` here?

**Comment by Oleh-Stefanovych** [comment]
Sorry, last 5 commits are mistake. They must be reverted.

File: VoteCalculator/app/model/vote-calculator/index.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+     * @returns {ElectionOutcome} - the result of election process. Contains chosen leaders and the
+     * number of votes that they received.
+     */
+    electLeaders() {
+
+        const electionCommission = new ElectionCommission(this._votingNodes);
+
+        Array.from(this._votingNodes).find((votingNode) => {
+            electionCommission.submitVoteFor(votingNode.voteForCandidate(this._votingNodes));
+            return electionCommission.isThereAnEarlyLeader;
```
Comment (Oleh-Stefanovych) [inline VoteCalculator/app/model/vote-calculator/index.js]: We discussed about `includes(...)` method, but it has no a callback function.

**Comment by Oleh-Stefanovych** [comment]
@dkalinin, today`s requirements are met, PTAL.

File: VoteCalculator/app/model/vote-calculator/index.js (hunk 8 above)
Comment (Oleh-Stefanovych) [reply VoteCalculator/app/model/vote-calculator/index.js]: Instead of `find(...)`

File: VoteCalculator/app/model/vote-calculator/index.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+     *
+     * @returns {ElectionOutcome} - the result of election process. Contains chosen leaders and the
+     * number of votes that they received.
+     */
+    electLeaders() {
+
+        const electionCommission = new ElectionCommission(this._votingNodes);
+
+        Array.from(this._votingNodes).find((votingNode) => {
+            electionCommission.submitVoteFor(votingNode.voteForCandidate(this._votingNodes));
```
Comment (dkalinin) [inline VoteCalculator/app/model/vote-calculator/index.js]: Too long construction. Please create a variable for `votingNode.voteForCandidate(this._votingNodes)`.

File: VoteCalculator/test/model/election-commission/index.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            .add(this.ariaNode)
+            .add(this.sansaNode);
+
+        this.electionCommission = new ElectionCommission(candidates);
+    }
+});
+
+QUnit.test('Test of votes submitting.', function (assert) {
+
+    this.electionCommission.submitVoteFor(this.robNode);
```
Comment (dkalinin) [inline VoteCalculator/test/model/election-commission/index.js]: Create a local variable for `this.electionCommission`.

**Pull request merged**